lib/msal-core/src/UserAgentApplication.ts

```diff
@@ -1111,18 +1111,21 @@ export class UserAgentApplication {
             // Redirect to home page if login request url is null (real null or the string null)
             if (!loginRequestUrl || loginRequestUrl === "null") {
                 this.logger.error("Unable to get valid login request url from cache, redirecting to home page");
-                window.location.href = "/";
+                window.location.assign("/");
                 return;
-            } else if (currentUrl !== loginRequestUrl) {
-                // If loginRequestUrl contains a hash (e.g. Angular routing), process the hash now then redirect to prevent both hashes in url
-                if (loginRequestUrl.indexOf("#") > -1) {
-                    this.logger.info("loginRequestUrl contains hash, processing response hash immediately then redirecting");
-                    this.processCallBack(hash, stateInfo, null);
```

Comment:
I'm interested as to why we don't need to call `processCallback` here anymore, can you provide more detail here?

Reply:
Yeah so before we processed the token hash and then redirected to the `loginStartPage` with, what we'll call, the user hash. Now instead redirection only occurs when host/path doesn't match. If it does we will only change the hash of the page to the user hash instead of doing a full redirect, and then process the token hash (which was extracted at the beginning of the function in both scenarios)

Reply:
Essentially I changed the logic to pull apart the `loginStartPage` into the different parts and make sure the final page matches, while also making sure we only redirect when we need to and the token hash gets processed on the final page

Reply:
Okay, so the process for this scenario (redirect start page includes a non-msal hash) is:

1. Redirect page loads with MSAL hash, navigateToLoginRequestUrl is true, loginRequestUrl is set with non-msal hash
2. MSAL hash is parsed and stored
3. Check to see if current url is different from loginRequest url, which would be false in this scenario since redirect uris cant contain hashes.
4. Since it is false, redirect to loginRequestUrl, which is where the stored MSAL hash is processed.

Correct?

Reply:
Not exactly. The flow would go more like this: 

1. Redirect page loads with MSAL hash, navigateToLoginRequestUrl is true, loginRequestUrl is set with non-msal hash
2. Strip msal hash, hold in local variable
3. Check if current url is different from loginRequest url, not taking into account any hash or query string param in either url
4. If the previous check is false, redirect to the loginRequestUrl (without non-msal hash/query string) + msal hash
5. Now step 3 is true, no more redirect required
6. Check cached loginRequestUrl for non-msal hash or query string. If present, add these onto the current window url
7. Process and store msal hash from step 2

Happy to go into this in more detail on Wednesday if needed.

Reply:
Ah okay. Didn't see the `processCallback` at the bottom of the function. Tested this with the Angular sample with hash routing and redirects, and worked as expected.